### ISSUE TITLE
PipeExtensions.WriteStreamedMessageAsync throws System.ObjectDisposedException: IFeatureCollection has been disposed.

### ISSUE THREAD

**Issue by kiwidev** [grpc/grpc-dotnet#1845]
### What version of gRPC and what language are you using?
grpc-dotnet 2.47

### What operating system (Linux, Windows,...) and version?
Linux - Mariner 2.0

### What runtime / compiler are you using (e.g. .NET Core SDK version `dotnet --info`)
.NET 6.0.400

### What did you do?
Under load scenarios, we've observed a new set of exceptions being logged when writes fail.

There was a recent change made to PipeExtensions (bfe312a3baeea3dc3676b3770dedd1b927e42fb3) which now references the serverCallContext.CancellationToken after the pipeWriter's flush fails.

It appears that if this write request occurs after the HTTP request has been disposed then one of the following exceptions can be thrown:
```
System.ObjectDisposedException: IFeatureCollection has been disposed.
Object name: 'Collection'.
   at Microsoft.AspNetCore.Http.Features.FeatureReferences`1.ThrowContextDisposed()
   at Microsoft.AspNetCore.Http.DefaultHttpContext.get_RequestAborted()
   at Grpc.AspNetCore.Server.Internal.PipeExtensions.WriteStreamedMessageAsync[TResponse](PipeWriter pipeWriter, TResponse response, HttpContextServerCallContext serverCallContext, Action`2 serializer, CancellationToken cancellationToken)
   at Grpc.AspNetCore.Server.Internal.HttpContextStreamWriter`1.WriteCoreAsync(TResponse message, CancellationToken cancellationToken)
```
```
System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.AspNetCore.Http.Features.FeatureReferences`1.UpdateCached[TFeature,TState](TFeature& cached, TState state, Func`2 factory, Int32 revision, Boolean flush)
   at Microsoft.AspNetCore.Http.DefaultHttpContext.get_RequestAborted()
   at Grpc.AspNetCore.Server.Internal.PipeExtensions.WriteStreamedMessageAsync[TResponse](PipeWriter pipeWriter, TResponse response, HttpContextServerCallContext serverCallContext, Action`2 serializer, CancellationToken cancellationToken)
   at Grpc.AspNetCore.Server.Internal.HttpContextStreamWriter`1.WriteCoreAsync(TResponse message, CancellationToken cancellationToken)
```

It's possible we can work around this via various locks to ensure that the request has not been disposed before sending messages (in this scenario the response was sent in response to a callback so not 100% synchronized with the gRPC server method completion), but I wonder if the code needs to read the `serverCallContext.CancellationToken` prior to calling Flush.

**Comment by JamesNK** [comment]
If `serverCallContext.CancellationToken` was accessed prior to flush then a `CancellationTokenSource` would always be allocated for the request (it's lazily created internally).

Even if the CT was accessed earlier, I imagine you'd still get the same error when trying to write after the request is done.

Accessing and using HttpContext after a request is done will always produce unpredictable results. Getting an error like this is the best case. Worse case is the HttpContext is quickly pooled and reused with another request and things get really messy.

**Comment by JamesNK** [comment]
I added some docs that discuss multi-threading problems: https://github.com/dotnet/AspNetCore.Docs/pull/26767

**Comment by kiwidev** [comment]
Nice work on the docs, and agree about everything you've stated about concurrency. 👍  I had raised this issue because it was a new error showing up inside our logs (that we initially thought was due to changes inside our code rather than framework).

Likely the root cause of our issue was caused by aborting the stream write in response to the ASP.Net cancellation token rather than waiting for the Write to fully complete.
